### DISCUSSION TITLE
What exactly is a floating window in clap?

### DISCUSSION THREAD

**Discussion by logos-maker** [free-audio/clap#142] [Q&A] [answered]
Is it the plugin that just makes it own independent window?
Or it the plugin dependent on taking over a window handle given by the host?

I would say that 99% percent of all crashes is that you don't make your own independent window. As the plugin creator doesn't know what the host does. And the host don't know what the plugin will do. And that's probably the one reason that audio plugins is related to the most caches in a modern computer without exception.

And to not have example code in C code for is like begging for problems. As C++ hides away what code actually does in layer upon layers in class definitions. So no one has a clear understanding of what thing is supposed to actually do. That's why all libraries with more installations than humans in the world is C only, and with C example code, and that's why for example the Linux kernel is C only. And that gives people the freedom to use any language, and not just a subset of languages that is heavily reliant on C++. 

C code and C code examples, is the best way to kill of ambiguities that comes from reliance in hiding away the truth with C++. And with C++ examples only,  people with end up in superstition about what things should be interpreted as, when different hosts and plugins start to crash each other.

**Comment by baconpaul** [comment, marked as answer]
the documentation in include/clap/ext/gui.h answer your question but as you surmised, floating means the plugin makes its own window and set parent is never called, and non floating means the host makes a window and set parent is called. In the floating case set_transient may be called for window ordering hints.

**Comment by robbert-vdh** [comment]
An example use case for these floating windows is Wine plugin bridging. Wine windows don't like being embedded in other windows and you need a ton of hacks to make it work, so if both the host and the plugin supports these freestanding windows then you can skip the whole embedding part altogether for those plugins. Of course, hosts like REAPER that embed the plugin's editor inside of a window with more controls are unlikely to support these freestanding windows.

**Comment by baconpaul** [comment]
Do you know if any hosts support floating right now @robbert-vdh ? (I know none of the juce plugins implement it, nor does the clap saw demo).

**Reply by robbert-vdh** [reply]
Bitwig might, I have not checked.

**Reply by abique** [reply]
Bitwig does.

**Reply by logos-maker** [reply]
I made one and tested it in Bitwig, it works. But It handles that Icon for opening and closing the plug GUI in an odd way. I think it may be a residue from how they mange VST's. As it resembles how it acts when I've been trying to make a VST2.

After I have been crashing the host hundreds of times, I know that sand-boxing of plugins is not fail-safe. You can still get to the host and mess it up. And you have to restart the host to be sure it's still acts the right way.

Now I'm eventually passed that stage. But in fresh remembrance on how you can mess up the host even with so called sand boxing. For real sand-boxing, I guess the host has to mimic what the OS is doing without getting it involved. It's probably more doable in Linux, as it's possible to make your own x11 sever running inside your own host, using xephyr as an example. But that's still a bit insane stretch to protect the host.
